substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaLangSubstitutions.java

```diff
@@ -722,6 +745,348 @@ static Package getPackage(Class<?> c) {
+        final StackIterator iter = new StackIterator(sp, ip, anchor, options);
+        try {
+            // todo: it would be nice if Stream.iterator() just returned our *original* iterator
+            return function.apply(StreamSupport.stream(Spliterators.spliterator(iter, 0L, Spliterator.NONNULL), false));
+        } finally {
+            iter.invalidate();
+        }
+    }
+
+    @Substitute
+    public Class<?> getCallerClass() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Substitute
+    public static Target_java_lang_StackWalker getInstance() {
+        return StackWalkerHolder.PLAIN;
+    }
+
+    @Substitute
+    public static Target_java_lang_StackWalker getInstance(Option option) {
+        if (option == Option.RETAIN_CLASS_REFERENCE) {
+            throw new UnsupportedOperationException(option.toString());
+        } else if (option == Option.SHOW_REFLECT_FRAMES) {
+            return StackWalkerHolder.SRF;
+        } else if (option == Option.SHOW_HIDDEN_FRAMES) {
+            return StackWalkerHolder.SHF;
+        } else {
+            return StackWalkerHolder.PLAIN;
+        }
+    }
+
+    @Substitute
+    public static Target_java_lang_StackWalker getInstance(Set<Option> option) {
+        if (option.contains(Option.RETAIN_CLASS_REFERENCE)) {
+            throw new UnsupportedOperationException(option.toString());
+        } else if (option.contains(Option.SHOW_REFLECT_FRAMES)) {
+            return StackWalkerHolder.SRF;
+        } else if (option.contains(Option.SHOW_HIDDEN_FRAMES)) {
+            return StackWalkerHolder.SHF;
+        } else {
+            return StackWalkerHolder.PLAIN;
+        }
+    }
+
+    @Substitute
+    public static Target_java_lang_StackWalker getInstance(Set<Option> option, int estDepth) {
+        return getInstance(option);
+    }
+}
+
+final class StackIterator implements Iterator<Target_java_lang_StackWalker.StackFrame> {
+    final Thread thread = Thread.currentThread();
+    final boolean includeReflect;
+    final boolean includeHidden;
+    boolean valid = true;
+    JavaFrameAnchor anchor;
+    Pointer sp;
+    CodePointer ip;
+    Target_java_lang_StackWalker.StackFrame next;
+    Iterator<Target_java_lang_StackWalker.StackFrame> nested;
+
+    StackIterator(final Pointer sp, final CodePointer ip, final JavaFrameAnchor anchor, final Set<Target_java_lang_StackWalker.Option> options) {
+        includeHidden = options.contains(Target_java_lang_StackWalker.Option.SHOW_HIDDEN_FRAMES);
+        includeReflect = includeHidden || options.contains(Target_java_lang_StackWalker.Option.SHOW_REFLECT_FRAMES);
+        this.sp = sp;
+        this.ip = ip;
+        this.anchor = anchor;
+    }
+
+    public boolean hasNext() {
+        checkThread();
+        while (next == null) {
+            while (nested == null) {
+                while (anchor.isNonNull() && anchor.getLastJavaSP().belowOrEqual(sp)) {
```

Comment:
Why do you need to duplicate the code from `JavaStackWalker` here? `JavaStackWalker` is already a generic mechanism where you can plug in any stack frame visitor you want to write.

Reply:
Because the access pattern of the JDK `StackWalker` is fundamentally incompatible with this model from a performance standpoint.  It does not call back to user code for each frame; rather, the user can draw out each frame via a Stream or Iterator.  If I used the `JavaStackWalker` approach, I would have to iterate the entire stack into an array or list, and then stream that to the user instead, or else do something that would result in a similar performance profile.  This would be unfortunate, because this stack walk otherwise has potential to be quite a bit more performant than HotSpot's one.

If there was a clean way to reuse code between the two, I'd be willing to do it, but I don't see such a way.

Reply:
OK, then we need a bigger refactoring of our stack walking code to avoid code duplication.

Reply:
OK.  I have... maybe... an idea.  I'll give it a try and see how it goes.